components/experimental/src/duration/format.rs

```diff
@@ -0,0 +1,528 @@
+
+        let ffd = self.fmt.fdf.format(&fd);
+        sink.with_part(parts::HOUR, |e| ffd.write_to_parts(e))?;
+
+        // 12. Return result.
+        Ok(())
+    }
+
+    /// Section 1.1.10
+    /// Formats numeric minutes to sink. Requires minutes formatting style to be either Numeric or TwoDigit.
+    fn format_numeric_minutes<V: PartsWrite + ?Sized>(
+        &self,
+        minute_val: u64,
+        sign_displayed: &mut bool,
+        hours_formatted: bool,
+        sink: &mut V,
+    ) -> core::fmt::Result {
+        // 1. Let result be a new empty List.
+
+        // 2. If hoursDisplayed is true, then
+        if hours_formatted {
+            // a. Let separator be durationFormat.[[DigitalFormat]].[[HoursMinutesSeparator]].
+            // b. Append the Record { [[Type]]: "literal", [[Value]]: separator, [[Unit]]: empty } to result.
+            sink.with_part(parts::LITERAL, |e| {
+                e.write_str(self.fmt.digital.get().separator.as_ref())
+            })?;
+        }
+
+        // 3. Let minutesStyle be durationFormat.[[MinutesStyle]].
+        // 4. Assert: minutesStyle is "numeric" or minutesStyle is "2-digit".
+        debug_assert!(
+            self.fmt.options.minute == FieldStyle::Numeric
+                || self.fmt.options.minute == FieldStyle::TwoDigit
+        );
+
+        // 5. Let nfOpts be OrdinaryObjectCreate(null).
+        // 6. Let numberingSystem be durationFormat.[[NumberingSystem]].
+        // 7. Perform ! CreateDataPropertyOrThrow(nfOpts, "numberingSystem", numberingSystem).
+        let mut fd = FixedDecimal::from(minute_val);
+
+        // 8. If minutesStyle is "2-digit", then
+        if self.fmt.options.minute == FieldStyle::TwoDigit {
+            // a. Perform ! CreateDataPropertyOrThrow(nfOpts, "minimumIntegerDigits", 2𝔽).
+            fd.pad_start(2);
+        }
+
+        // 9. If signDisplayed is false, then
+        if !*sign_displayed {
+            // a. Perform ! CreateDataPropertyOrThrow(nfOpts, "signDisplay", "never").
+            fd.apply_sign_display(SignDisplay::Never);
+        }
+
+        // 10. Let nf be ! Construct(%NumberFormat%, « durationFormat.[[Locale]], nfOpts »).
+        // 11. Let minutesParts be ! PartitionNumberPattern(nf, minutesValue).
+        // 12. For each Record { [[Type]], [[Value]] } part of minutesParts, do
+        // a. Append the Record { [[Type]]: part.[[Type]], [[Value]]: part.[[Value]], [[Unit]]: "minute" } to result.
+
+        let ffd = self.fmt.fdf.format(&fd);
+        sink.with_part(parts::MINUTE, |e| ffd.write_to_parts(e))?;
+
+        // 13. Return result.
+        Ok(())
+    }
+
+    /// Section 1.1.7
+    /// Computes the sum of all values in durationFormat units with "fractional" style,
+    /// expressed as a fraction of the smallest unit of durationFormat that does not use "fractional" style.
+    fn add_fractional_digits(&self, original: u64) -> FixedDecimal {
```

Comment:
Issue: This function can be cleaned up a lot to work all in FixedDecimal space, assuming that the input has the same constraints as it has in ECMAScript

Reply:
I found [`fixed_decimal::FixedDecimal::concatenate_end`](https://docs.rs/fixed_decimal/latest/fixed_decimal/struct.FixedDecimal.html#method.concatenate_end) while writing this, but it apparently fails for overlapping magnitudes, which is possible when, for example, milliseconds are larger than three digits.